tests/integration_tests/datasources/test_nocloud.py

```diff
@@ -88,3 +95,237 @@ def test_nocloud_seedfrom_vendordata(client: IntegrationInstance):
+        # environments, the ftp server would be separate from the instance.
+        #
+        # For a reliable server that fits with the framework of running tests
+        # on a single instance, it is easier to just install an ftp server
+        # that runs on the second boot prior to the cloud-init unit which
+        # reaches out to the ftp server. This achieves reaching out to an
+        # ftp(s) server for testing - cloud-init just doesn't have to reach
+        # very far to get what it needs.
+        #
+        # DO NOT use these concepts in a production.
+        #
+        # This configuration is neither secure nor production-grade - intended
+        # only for testing purposes.
+        client.write_to_file(
+            "/server.py",
+            dedent(
+                """\
+                #!/usr/bin/python3
+                import logging
+
+                from pyftpdlib.authorizers import DummyAuthorizer
+                from pyftpdlib.handlers import FTPHandler, TLS_FTPHandler
+                from pyftpdlib.servers import FTPServer
+                from pyftpdlib.filesystems import UnixFilesystem
+
+                encrypted = """
+                + str(encrypted)
+                + """
+
+                logging.basicConfig(level=logging.DEBUG)
+
+                # yeah, it's not secure but that's not the point
+                authorizer = DummyAuthorizer()
+
+                # Define a read-only anonymous user
+                authorizer.add_anonymous("/home/anonymous")
+
+                # Instantiate FTP handler class
+                if not encrypted:
+                    handler = FTPHandler
+                    logging.info("Running unencrypted ftp server")
+                else:
+                    handler = TLS_FTPHandler
+                    handler.certfile = "/cert.pem"
+                    handler.keyfile = "/key.pem"
+                    logging.info("Running encrypted ftp server")
+
+                handler.authorizer = authorizer
+                handler.abstracted_fs = UnixFilesystem
+                server = FTPServer(("localhost", 2121), handler)
+
+                # start the ftp server
+                server.serve_forever()
+                """
+            ),
+        )
+        assert client.execute("chmod +x /server.py").ok
+
+        if encrypted:
+            if CURRENT_RELEASE > FOCAL:
+                assert client.execute("apt install -yq mkcert").ok
+            else:
+
+                # install golang
+                assert client.execute("apt install -yq golang").ok
+
+                # build mkcert from source
+                #
+                # we could check out a tag, but the project hasn't
+                # been updated in 2 years
+                #
+                # instructions from https://github.com/FiloSottile/mkcert
+                assert client.execute(
+                    "git clone https://github.com/FiloSottile/mkcert && "
```

Comment:
This changes where this can be run, and how safe it is to run this test. Could we package a fixed cert.pem and key.pem (set to expire in a hundred years or so?) to skip this download, compile, and execute?

Reply:
> Could we package a fixed cert.pem and key.pem (set to expire in a hundred years or so?) to skip this download, compile, and execute?

Given that this will be dropped as soon as we are no longer releasing to focal, I'm not sure that it is worth it. Also, I could use help with cert generation for that. 

The challenging bit wasn't actually getting a cert.pem and key.pem, it was building a root cert and getting it into the CA store so that the client would trust it - self signed certs won't work for that. I'm not very well versed in the openssl commandline, but if someone is able to generate a test root CA, cert, and key I'd be happy to use that rather than using mkcert.

Reply:
"no longer releasing to focal" may be a very long way away.

The security team has some rough-and-ready python for testing OpenSSL that might be suitable for stealing:

https://git.launchpad.net/qa-regression-testing/tree/scripts/test-openssl.py#n238